File: front-end/src/services/bookService.js
import axios from 'axios'

const API = 'http://localhost:8080/api/v1/books'

export const getBooks = () => {
  return axios.get(API)
}

export const getAvailableBooks = () => {
  return axios.get(`${API}/available`)
}

export const getBookById = (bookId) => {
  return axios.get(`${API}/${bookId}`)
}

export const searchBooks = (keyword) => {
  return axios.get(`${API}/search?keyword=${keyword}`)
}

export const addBook = (bookData) => {
  return axios.post(API, bookData)
}

export const updateBook = (bookId, bookData) => {
  return axios.put(`${API}/${bookId}`, bookData)
}

export const deleteBook = (bookId) => {
  return axios.delete(`${API}/${bookId}`)
}

File: server/src/main/java/com/library/server/services/BookService.java
package com.library.server.services;

import com.library.server.models.Book;
import com.library.server.repositories.BookRepository;

import lombok.RequiredArgsConstructor;

import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
public class BookService {

    private final BookRepository bookRepository;

    // Add Book
    public Book addBook(Book book) {

        // Check duplicate ISBN
        if (bookRepository.findByIsbn(book.getIsbn()) != null) {
            throw new IllegalArgumentException("Book with ISBN already exists");
        }

        book.setAvailable(true);

        return bookRepository.save(book);
    }

    // Get All Books
    public List<Book> getAllBooks() {
        return bookRepository.findAll();
    }

    // Get Available Books
    public List<Book> getAvailableBooks() {
        return bookRepository.findByAvailableTrue();
    }

    // Search Books
    public List<Book> searchBooks(String title, String author) {

        return bookRepository
                .findByTitleContainingIgnoreCaseOrAuthorContainingIgnoreCase(
                        title,
                        author
                );
    }

    // Get Book By ID
    public Book getBookById(Long id) {

        return bookRepository.findById(id)
                .orElseThrow(() ->
                        new IllegalArgumentException("Book not found"));
    }

    // Update Book
    public Book updateBook(Book book) {
        return bookRepository.save(book);
    }

    // Delete Book
    public void deleteBook(Long id) {

        Book book = getBookById(id);

        bookRepository.delete(book);
    }
}